File: exonum-java-binding/core/src/main/java/com/exonum/binding/core/storage/database/ReadonlyFork.java
/*
 * Copyright 2020 The Exonum Team
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.exonum.binding.core.storage.database;

import static com.google.common.base.Preconditions.checkNotNull;

import com.exonum.binding.core.proxy.Cleaner;
import com.exonum.binding.core.proxy.NativeHandle;
import com.exonum.binding.core.proxy.ProxyDestructor;
import com.exonum.binding.core.util.LibraryLoader;
import com.google.common.annotations.VisibleForTesting;

/**
 * A readonly {@link Fork}-based database Access object.
 *
 * <p>A readonly Fork sees all the changes made to the base Fork, but forbids
 * modifications. That allows safely passing the readonly fork to foreign code.
 *
 * <p>A readonly fork forbids accessing the same index that is being modified in the base
 * fork. Also, accessing an index in the readonly fork will make it inaccessible in the base fork.
 */
public final class ReadonlyFork extends AbstractAccess {

  static {
    LibraryLoader.load();
  }

  private final Cleaner cleaner;

  private ReadonlyFork(NativeHandle nativeHandle, Cleaner cleaner) {
    super(nativeHandle, false);
    this.cleaner = cleaner;
  }

  /**
   * Creates a new ReadonlyFork from the native handle. The destructor will be registered
   * in the given cleaner.
   *
   * @param nativeHandle a handle of the native ReadonlyFork object
   * @param cleaner a cleaner to destroy the native peer and any dependent objects
   */
  public static ReadonlyFork fromHandle(long nativeHandle, Cleaner cleaner) {
    checkNotNull(cleaner);
    return fromHandleInternal(nativeHandle, cleaner);
  }

  @VisibleForTesting static ReadonlyFork fromFork(Fork fork) {
    long baseForkHandle = fork.getAccessNativeHandle();
    long roForkHandle = nativeCreate(baseForkHandle);
    return fromHandleInternal(roForkHandle, fork.getCleaner());
  }

  /**
   * Creates a ReadonlyFork from the Fork (basically, {@code Fork#readonly}.
   */
  private static native long nativeCreate(long baseForkHandle);

  /**
   * Expects validated parameters so that it does not throw and registers the destructor
   * properly, which is *required* to prevent leaks.
   */
  private static ReadonlyFork fromHandleInternal(long roForkNativeHandle, Cleaner cleaner) {
    NativeHandle handle = new NativeHandle(roForkNativeHandle);
    ProxyDestructor.newRegistered(cleaner, handle, ReadonlyFork.class,
        /* todo [ECR-4167]: Verify after native implementation! */ Accesses::nativeFree);
    return new ReadonlyFork(handle, cleaner);
  }

  @Override
  public Cleaner getCleaner() {
    return cleaner;
  }
}

File: exonum-java-binding/core/rust/src/storage/readonly_fork.rs
// Copyright 2018 The Exonum Team
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use exonum_merkledb::generic::{ErasedAccess, GenericAccess, GenericRawAccess};
use jni::{objects::JObject, JNIEnv};

use std::panic;

use {handle, into_generic_raw_access, utils::unwrap_exc_or_default, Handle};

/// Creates `ReadonlyFork` from base access (passed as `access_handle`).
///
/// Throws exception and returns null if creation of `ReadonlyFork` is not possible.
#[no_mangle]
pub extern "system" fn Java_com_exonum_binding_core_storage_database_ReadonlyFork_nativeCreate(
    env: JNIEnv,
    _: JObject,
    access_handle: Handle,
) -> Handle {
    let res = panic::catch_unwind(|| {
        let access = handle::cast_handle::<ErasedAccess>(access_handle);
        let readonly_fork = match access {
            GenericAccess::Raw(raw) => match raw {
                GenericRawAccess::Fork(fork) => fork.readonly(),
                GenericRawAccess::OwnedFork(fork) => fork.readonly(),
                _ => panic!("Attempt to create ReadonlyFork from non-fork: {:?}", access),
            },
            _ => panic!("Attempt to create ReadonlyFork from non-Fork: {:?}", access),
        };
        let readonly_fork = ErasedAccess::from(unsafe { into_generic_raw_access(readonly_fork) });
        let handle = handle::to_handle(readonly_fork);
        Ok(handle)
    });

    unwrap_exc_or_default(&env, res)
}
